torch/library.py

```diff
@@ -0,0 +1,87 @@
+from ._ops import OpOverload
+from typing import Set
+import traceback
+import torch._C as C
+__all__ = ['extend_library']
+
+# User created libraries to extend existing libraries
+# Each user created library is added here to ensure that it's not automatically removed outside the
+# scope of the function it was created in.
+_impls_for_existing_libraries = {}
+
+# Set containing the combination of (namespace, operator, DispatchKey) for which a new kernel has been registered
+# The keys in the set are of the form `namespace + "/" + op_name + "/" + dispatch_key`.
+# This set is maintained to ensure that two libraries don't try to override the exact same functionality to avoid
+# libraries calling into kernels not intended to be called.
+_impls: Set[str] = set()
+
+class _Library:
+    # kind can be DEF, IMPL
```

Comment:
Union?
Also I would expect the check below if the kind if valid to happen before we create the c++ object

Reply:
yes that's what we do

Reply:
Isn;t the init function doing:
```python
        # stuff
        self.m = C._dispatch_library(kind, ns, dispatch_key, filename, lineno)
        # more stuff
        if kind != "IMPL":
            raise ValueError("Unsupported kind: ", kind)
```

So that is checked after?